repose-aggregator/components/filters/tenant-culling-filter/src/main/scala/org/openrepose/filters/tenantculling/TenantCullingFilter.scala

```diff
@@ -48,61 +45,41 @@ import org.slf4j.{Logger, LoggerFactory}
-    val relevantRoles = request.getSplittableHeaderScala(RELEVANT_ROLES)
-    if (cacheKey != null) {
-      try {
-        val token = objectSerializer.readObject(objectSerializer.writeObject(datastore.get(cacheKey))).asInstanceOf[KeystoneV2Common.ValidToken]
-        if (token != null) {
-          logNamedSeq("Token Roles", token.roles)
-          logNamedSeq("Relevant Roles", relevantRoles)
-          val tenants: Seq[String] = token.roles
-            .filter(role => relevantRoles.contains(role.name))
-            .filter(_.tenantId.isDefined)
-            .map(_.tenantId.get)
-          request.removeHeader(TENANT_ID)
-          val withDefault = withDefaultTenant(token, tenants)
-          log.debug("Adding {}:", TENANT_ID)
-          withDefault.foreach { tenant =>
-            log.debug(" - {}", tenant)
-            request.addHeader(TENANT_ID, tenant)
-          }
+    val rolesMap = Option(request.getHeader(TENANT_ROLES_MAP))
+    val relevantRoles = request.getSplittableHeaderScala(RELEVANT_ROLES).map(_.split('/').head)
+    rolesMap match {
+      case Some(tenantToRolesMap) =>
+        try {
+          val culledMap = unpackMap(tenantToRolesMap).filter({ case(tenant, roles) => roles.toList.intersect(relevantRoles).nonEmpty })
+          request.replaceHeader(TENANT_ROLES_MAP, packMap(culledMap))
+          request.replaceHeader(TENANT_ID, culledMap.keySet.filterNot("repose/domain/roles".equals).mkString(","))
           chain.doFilter(request, response)
+        } catch {
+          case e@(_: IllegalArgumentException | _: JsonParseException) =>
+            logger.error("A problem occurred while trying to parse the role map.", e)
+            response.sendError(SC_INTERNAL_SERVER_ERROR)
         }
-        else {
-          log.debug("Cache miss for key: {}", cacheKey)
-          response.sendError(SC_UNAUTHORIZED)
-        }
-      } catch {
-        case cnfe: ClassNotFoundException =>
-          log.error("This shouldn't have been possible, somehow the item that came back from datastore doesn't match a class available in the current classloader", cnfe)
-          response.sendError(SC_INTERNAL_SERVER_ERROR)
-      }
-    }
-    else {
-      log.debug("Cache key header not found")
-      response.sendError(SC_UNAUTHORIZED)
+      case None =>
+        logger.debug("Tenant to roles map header not found")
+        response.sendError(SC_INTERNAL_SERVER_ERROR)
     }
   }
 
   @Override override def destroy(): Unit = {
   }
 }
+
+object TenantCullingFilter {
+  type TenantToRolesMap = Map[String, Set[String]]
+
+  def packMap(tenantToRolesMap: TenantToRolesMap): String = {
+    val tenantToRolesJson = Json.stringify(Json.toJson(tenantToRolesMap))
+    Base64.getEncoder.encodeToString(tenantToRolesJson.getBytes)
+  }
+
+  def unpackMap(tenantToRolesMap: String): TenantToRolesMap = {
+    Json.parse(new String(Base64.getDecoder.decode(tenantToRolesMap))).as[TenantToRolesMap]
```

Comment:
These Base64 En/Decoders can be replaced by using the methods in:
> `org.openrepose.commons.utils.string.Base64Helper`

Reply:
Yep, they didn't exist when this was originally done. I'll swap.